bolt-modules/boltlib/lib/puppet/functions/resource.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# Lookup a resource in the target's data.
+#
+# For more information about resources see [the
+# documentation](https://puppet.com/docs/puppet/latest/lang_resources.html).
+#
+# > **Note:** The `ResourceInstance` data type is under active development and is subject to
+#   change. You can read more about the data type in the [experimental features
+#   documentation](experimental_features.md#resourceinstance-data-type).
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:resource) do
+  # Lookup a resource in the target's data.
+  # @param target The `Target` object to add resources to. See {get_targets}.
+  # @param type The type of the resource
+  # @param title The title of the resource
+  # @return The ResourceInstance if found, or Undef
+  # @example Get the openssl package resource
+  #   $target.apply_prep
+  #   $resources = $target.get_resources(Package).first['resources']
+  #   $target.set_resources($resources)
+  #   $openssl = $target.resource('Package', 'openssl')
+  dispatch :resource do
+    param 'Boltlib::TargetSpec', :target
```

Comment:
```suggestion
    param 'Target', :target
```

This can probably just be a Target object since one already has to exist to even retrieve resources. This would also be in line with the `set_resources` function, which expects a Target object.

Reply:
This just limits how users can call the function. We call `get_target` on the parameter, so with `Boltlib::TargetSpec` users can do `resource('hostname', 'type', 'title')` where with just `Target` that will fail. I don't see a good reason to limit it.

Reply:
For consistency with `facts()`, `vars()`, and `resources()`, I think it makes sense for this to be `Target`.